course/models.py

```diff
@@ -1587,6 +1587,23 @@ def get_state_desc(self):
         return dict(GRADE_STATE_CHANGE_CHOICES).get(
                 self.state)
 
+    # Compare GradeChange objects in terms of time, in order to correctly sort
+    # objects, fixing #236 and #417
+    def __eq__(self, other):
+        if not (self.flow_session and other.flow_session):
+            # at least one object has flow_session, then compare them by grade_time
+            return self.grade_time == other.grade_time
+        else:
+            # if both objects have a flow_session
+            return self.flow_session.completion_time == other.completion_time
+
+    def __gt__(self, other):
+        if not (self.flow_session and other.flow_session):
+            return self.grade_time > other.grade_time
+        else:
+            return self.flow_session.completion_time > other.completion_time
```

Comment:
I think this might not yield a consistent ordering in the presence of grade changes with and without flow sessions.

Better:

```
self_time = self.grade_time
if self.flow_session is not None:
    self_time = self.flow_session.completion_time
other_time = ...
```

Same for `__eq__`. Maybe this time getting should be bagged up in its own function. There is also a potential performance concern, since `self.flow_session.completion_time` involves another query unless the flow sessions were brought in with `select_related` or some such. So we should comb through the usage sites of this and make sure that's the case.

Reply:
My idea was the compare should be determined by which objects we are compare. (Selecting key according to which type of Gradechange objects we are comparing). 

1. If either objects have no flow_session, we use grade_time as key.

1. If both have flow_session, we use flow_session.completion_time as key.

I don't think  `self_time` can achieve that, because you don't know yet who is going to be compared.

Reply:
> There is also a potential performance concern, since self.flow_session.completion_time involves another query unless the flow sessions were brought in with select_related or some such. So we should comb through the usage sites of this and make sure that's the case.

That's a good point.

Reply:
> I think this might not yield a consistent ordering in the presence of grade changes with and without flow sessions.

Can you give an example?

About inconsistency might happen if two GradeChange objects have the same time attribute (we selected, though it's rare) while with different percentage value. So I'll add pk if that happens.

Reply:
(Then `__eq__` is meaningless.)

Reply:
> Can you give an example?

I'll try:
```
A: grade time 2015
B: grade time 2016, flow completed 2014
C: grade time 2017
D: grade time 2018, flow completed 2013
```

Then `A < B < C`

But also `D < B` and `C < D`, which is a direct contradiction.

Reply:
That's interesting. I'll think about that tomorrow.

Reply:
Since that inconsistency exists (as you had demostrated), defining `__gt__` is not a good work around, at least it might be wrongly used in the future, I think it's better to implement the sorting within the range of `consume` method.

Reply:
Both previous commits will fail when reopen a graded (finished session). Because those GradeChange objects has a flow_session, but that session don't have a completion_time now.

Reply:
> Since that inconsistency exists (as you had demostrated), defining `__gt__` is not a good workaround, at least it might be wrongly used in the future

It's perfectly possible to define `__gt__` in a manner that doesn't run into inconsistencies. See my first comment with the `self_time`.

I guess I'll spell out the full set of methods that I think makes sense:

```python
def sort_time(self):
    return self.flow_session.completion_time if self.flow_session is not None else self.grade_time

def __lt__(self, other):
    if not isinstance(other, GradeChange):
        return NotImplemented

    return self.sort_time() < other.sort_time()

def __le__(self):
    if not isinstance(other, GradeChange):
        return NotImplemented

    return self.sort_time() <= other.sort_time()

def __gt__(self):
    if not isinstance(other, GradeChange):
        return NotImplemented

    return self.sort_time() > other.sort_time()

def __ge__(self):
    if not isinstance(other, GradeChange):
        return NotImplemented

    return self.sort_time() >= other.sort_time()
```

Reply:
That sounds good. I have finally done with my grading work of the ended semester. And I'll look into this today.

Reply:
For this part
```python
def sort_time(self):
    return self.flow_session.completion_time if self.flow_session is not None else self.grade_time
```

For some GradeChange objects, it does have a `flow_session`, but when that session is reopened, then `completion_time` is None, which is not comparable and the sorting will fail.

Reply:
And, we should not implement a method which can give an equal compare. Because that's another source of inconsistency. so `__ge__` and `__le__` should be removed. When `sort_time()` return two equal result, we should use only the one which comes later, IOW, which has greater `pk`.

Reply:
> For some GradeChange objects, it does have a flow_session, but when that session is reopened, then completion_time is None, which is not comparable and the sorting will fail.

Fair. So, catch that, and treat that case as if no flow session existed?

Reply:
>  so __ge__ and __le__ should be removed.

No objection. What methods does Python's sorting use?

Reply:
> No objection. What methods does Python's sorting use?

Simply `gt` might be enough.

Reply:
You know, I think you're right. If we can't supply a full suite of comparison operators, then maybe this is better handled by a `key=` function handed to `sort`.

Reply:
As for you example:
```
A: grade time 2015
B: grade time 2016, flow completed 2014
C: grade time 2017
D: grade time 2018, flow completed 2013
```
If `use_earliest`, then A will be used.
If `use_latest`, then C will be used.

This is tested in the latest commit. The logic is simpler, just keep the GradeChange object with the latest completed session when sorting. This mean that we need to use the grading strategy when consuming.